File: Activities/ArticleListActivity.cs
using System;
using System.Collections.Generic;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Support.V4.Content;
using Android.Views;
using Android.Widget;
using static Android.Widget.AdapterView;

namespace VorratsUebersicht
{
    [Activity(Label = "@string/Main_Button_ArtikelListe", Icon = "@drawable/ic_local_offer_white_48dp")]
    public class ArticleListActivity : Activity, SearchView.IOnQueryTextListener
    {
        List<ArticleListView> liste = new List<ArticleListView>();
        private IParcelable listViewState;

		private bool selectArticleOnly;
        private string category;
        private string subCategory;
        private bool   withoutCategory;
        private bool   notInStorage;
        private bool   notInShoppingList;
        private string eanCode;
        private string lastSearchText = string.Empty;
        private int    specialFilter = 0;
        private List<string> categoryList;

        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);

            this.selectArticleOnly = Intent.GetBooleanExtra("SelectArticleOnly", false);

            this.category          = Intent.GetStringExtra ("Category") ?? string.Empty;
            this.subCategory       = Intent.GetStringExtra ("SubCategory") ?? string.Empty;
            this.notInStorage      = Intent.GetBooleanExtra("NotInStorage", false);
            this.notInShoppingList = Intent.GetBooleanExtra("NotInShoppingList", false);
            this.eanCode           = Intent.GetStringExtra ("EANCode") ?? string.Empty;

            if (!string.IsNullOrEmpty(this.subCategory))
            {
                this.Title = string.Format("{0} - {1}", this.Title, this.subCategory);
            }

            SetContentView(Resource.Layout.ArticleList);

            // ActionBar Hintergrund Farbe setzen
            var backgroundPaint = ContextCompat.GetDrawable(this, Resource.Color.Application_ActionBar_Background);
            backgroundPaint.SetBounds(0, 0, 10, 10);
            ActionBar.SetBackgroundDrawable(backgroundPaint);
            ActionBar.SetDisplayHomeAsUpEnabled(true);

            this.categoryList = new List<string>();
            this.categoryList.Add(Resources.GetString(Resource.String.ArticleList_AllCategories));
            this.categoryList.Add(Resources.GetString(Resource.String.ArticleList_NoCategories));
            this.categoryList.AddRange(Database.GetCategoryAndSubCategoryNames());

            if (categoryList.Count > 1)
            {
                var categorySelection = FindViewById<LinearLayout>(Resource.Id.ArticleList_SelectCategory);
                categorySelection.Visibility = ViewStates.Visible;

                var spinnerCategory = FindViewById<Spinner>(Resource.Id.ArticleList_Categories);
                
                ArrayAdapter<String> dataAdapter = new ArrayAdapter<String>(this, Resource.Layout.Spinner_Black, this.categoryList);
                dataAdapter.SetDropDownViewResource(Android.Resource.Layout.SimpleSpinnerDropDownItem);
                spinnerCategory.Adapter = dataAdapter;

                spinnerCategory.ItemSelected += SpinnerCategory_ItemSelected;
            }

            ListView listView = FindViewById<ListView>(Resource.Id.ArticleList);
            listView.ItemClick += OnOpenArticleDetails;

            TextView articleListFilter = FindViewById<TextView>(Resource.Id.ArticleList_Filter);
            articleListFilter.Click += ArticleListFilter_Click;

            ImageView imageView = FindViewById<ImageView>(Resource.Id.ArticleList_FilterClear);
            imageView.Click += ArticleFilterClear_Click;

            ImageButton addButton = FindViewById<ImageButton>(Resource.Id.ArticleList_AddPosition);
            addButton.Click += AddArticle_Click;

            this.RegisterForContextMenu(listView);

            ShowArticleList();
        }

        private void ArticleListFilter_Click(object sender, EventArgs e)
        {
            this.FilterArticleList();
        }

        private void ArticleFilterClear_Click(object sender, EventArgs e)
        {
            this.specialFilter = 0;
            this.FindViewById<FrameLayout>(Resource.Id.ArticleList_FilterBanner).Visibility = ViewStates.Gone;

            this.ShowArticleList();
        }

        private void AddArticle_Click(object sender, EventArgs e)
        {
            // Create New Article
            this.ShowArticleDetails(0, null);
        }

        public override void OnCreateContextMenu(IContextMenu menu, View view, IContextMenuContextMenuInfo menuInfo)
        {
            if (view.Id == Resource.Id.ArticleList) 
            {
                menu.Add(IMenu.None, 1, IMenu.None, Resource.String.ArticleList_Lagerbestand);    // Lagerbestand
                menu.Add(IMenu.None, 2, IMenu.None, Resource.String.ArticleList_ToShoppingList);  // Auf Einkaufszettel
            }
        }

        public override bool OnContextItemSelected(IMenuItem item)
        {
            AdapterContextMenuInfo info = (AdapterContextMenuInfo) item.MenuInfo;

            ListView listView = FindViewById<ListView>(Resource.Id.ArticleList);
            ArticleListView  selectedItem = Tools.Cast<ArticleListView>(listView.Adapter.GetItem(info.Position));
            
            switch(item.ItemId)
            {
                case 1: // Lagerbestand
                    var storageDetails = new Intent(this, typeof(StorageItemQuantityActivity));
                    storageDetails.PutExtra("ArticleId", selectedItem.ArticleId);

                    this.SaveListState();
                    this.StartActivityForResult(storageDetails, 20);

                    return true;

                case 2: // Auf Einkaufsliste
                    AddToShoppingListDialog.ShowDialog(
                        this,
                        selectedItem.ArticleId,
                        null, null, this.RefreshArticleList);

                    return true;

                default:
                    return base.OnContextItemSelected(item);
            }
        }

        private void SpinnerCategory_ItemSelected(object sender, AdapterView.ItemSelectedEventArgs e)
        {
            string newCategoryName    = string.Empty;
            string newSubCategoryName = string.Empty;
            bool   withoutCategory    = false;

            string name = this.categoryList[e.Position];

            if (e.Position == 1)
            {
                withoutCategory = true;
            }

            if (e.Position > 1)
            {
                if (name.StartsWith("  - "))    // Ist das ein SubCategory?
                {
                    name = name.Substring(4);  // Mach aus "  - Gulasch" ein "Gulasch"
                    newSubCategoryName = name;
                }
                else
                {
                    newCategoryName = name;
                }
            }

            if ((newCategoryName != this.category) || (newSubCategoryName != this.subCategory) || withoutCategory != this.withoutCategory)
            {
                this.category        = newCategoryName;
                this.subCategory     = newSubCategoryName;
                this.withoutCategory = withoutCategory;

                this.ShowArticleList(this.lastSearchText);
            }
        }

        public override bool OnCreateOptionsMenu(IMenu menu)
        {
            MenuInflater.Inflate(Resource.Menu.ArticleList_menu, menu);

            var searchMenuItem = menu.FindItem(Resource.Id.ArticleList_Menu_Search);
            var searchView = (SearchView) searchMenuItem.ActionView;

            // https://coderwall.com/p/zpwrsg/add-search-function-to-list-view-in-android
            searchView.SetOnQueryTextListener(this);

            return base.OnCreateOptionsMenu(menu);
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            switch (item.ItemId)
            {
                case Android.Resource.Id.Home:
                    this.Finish();
                    return true;

                case Resource.Id.ArticleList_Menu_Add:
                    // Create New Article
                    this.ShowArticleDetails(0, null);
                    return true;

                case Resource.Id.ArticleList_Menu_Filter:

                    this.FilterArticleList();
                    break;

                case Resource.Id.ArticleList_Menu_Share:
                    this.ShareList();
                    return true;
            }
            return true;
        }

        private void FilterArticleList()
        {
            string[] actions = Resources.GetStringArray(Resource.Array.ArticleListeSpecialFilter);

            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.SetItems(actions, (sender2, args) =>
            {
                this.specialFilter = args.Which + 1;

                string filterText = this.Resources.GetString(Resource.String.ArticleList_SpecialFilter);
                filterText = String.Format(filterText, actions[args.Which]);

                this.FindViewById<FrameLayout>(Resource.Id.ArticleList_FilterBanner).Visibility = ViewStates.Visible;
                this.FindViewById<TextView>(Resource.Id.ArticleList_Filter).Text = filterText;

                this.ShowArticleList();
            });
            builder.Show();
        }

        private void ShareList()
        {
            if (MainActivity.IsGooglePlayPreLaunchTestMode)
            {
                return;
            }

            string text = string.Empty;

            foreach(ArticleListView view in this.liste)
            {
                if (!string.IsNullOrEmpty(view.Heading))     text += view.Heading     + "\n";
                if (!string.IsNullOrEmpty(view.SubHeading))  text += view.SubHeading  + "\n";
                if (!string.IsNullOrEmpty(view.Notes))       text += view.Notes + "\n";
                text += "\n";
            }

            TextView footer = FindViewById<TextView>(Resource.Id.ArticleList_Footer);
            text += footer.Text;

            string subject = string.Format("{0} - {1} {2}",
                Resources.GetString(Resource.String.Main_Button_Artikelangaben),
                DateTime.Now.ToShortDateString(),
                DateTime.Now.ToShortTimeString());

            Intent intentsend = new Intent();
            intentsend.SetAction(Intent.ActionSend);
            intentsend.PutExtra(Intent.ExtraSubject, subject);
            intentsend.PutExtra(Intent.ExtraText, text);
            intentsend.SetType("text/plain");
            
            StartActivity(intentsend);
        }

        private void RefreshArticleList()
        {
            this.SaveListState();
            this.ShowArticleList(this.lastSearchText);
            this.RestoreListState();
        }

        private void ShowArticleList(string text = null)
        {
            this.liste = new List<ArticleListView>();

            var articleList = Database.GetArticleList(
                this.category,
                this.subCategory,
                this.eanCode,
                this.notInStorage,
                this.notInShoppingList,
                this.withoutCategory,
                this.specialFilter,
                text);

            foreach(Article article in articleList)
            {
                liste.Add(new ArticleListView(article, this.Resources));
            }

            ArticleListViewAdapter listAdapter = new ArticleListViewAdapter(this, liste);
            listAdapter.OptionMenu += ListAdapter_OptionMenu;

            ListView listView = FindViewById<ListView>(Resource.Id.ArticleList);
            listView.Adapter = listAdapter;

            string status;

            if (articleList.Count == 1)
                status = string.Format(this.Resources.GetString(Resource.String.ArticleListSummary_Position), articleList.Count);
            else
                status = string.Format(this.Resources.GetString(Resource.String.ArticleListSummary_Positions), articleList.Count);

            TextView footer = FindViewById<TextView>(Resource.Id.ArticleList_Footer);
            footer.Text = status;
        }

        private void ListAdapter_OptionMenu(object sender, EventArgs e)
        {
            this.OpenContextMenu((TextView)sender);
        }

        private void OnOpenArticleDetails(object sender, AdapterView.ItemClickEventArgs e)
        {
            Java.Lang.Object itemObject = ((ListView)sender).GetItemAtPosition(e.Position);

            ArticleListView item = Tools.Cast<ArticleListView>(itemObject);

			// Nur Artikelauswahl, keine Detailsbearbeitung
			if (this.selectArticleOnly)
			{
                Intent intent = new Intent();
				intent.PutExtra("Heading",   item.Heading);
				intent.PutExtra("ArticleId", item.ArticleId);

                this.SetResult(Result.Ok, intent);

                this.Finish();
				return;
			}

            this.ShowArticleDetails(item.ArticleId, item.Heading);
        }

        private void SaveListState()
        {
            ListView listView = FindViewById<ListView>(Resource.Id.ArticleList);
            this.listViewState = listView.OnSaveInstanceState();
        }

        private void RestoreListState()
        {
            ListView listView = FindViewById<ListView>(Resource.Id.ArticleList);
            listView?.OnRestoreInstanceState(this.listViewState);
        }

        private void ShowArticleDetails(int articleId, string name)
        {
            var articleDetails = new Intent (this, typeof(ArticleDetailsActivity));
            articleDetails.PutExtra("Name", name);
            articleDetails.PutExtra("ArticleId", articleId);

            // Zum voranstellen
            articleDetails.PutExtra("Category",    this.category);
            articleDetails.PutExtra("SubCategory", this.subCategory);

            StartActivityForResult(articleDetails, 10);

            this.SaveListState();
        }

        protected override void OnActivityResult(int requestCode, [GeneratedEnum] Result resultCode, Intent data)
        {
            if (resultCode == Result.Ok)
            {
                this.ShowArticleList(this.lastSearchText);

                if (this.listViewState == null)
                    return;

                this.RestoreListState();    
            }
        }

        public bool OnQueryTextChange(string filter)
        {
            if (this.lastSearchText == filter)
                return true;

            // Filter ggf. mit Adapter, siehe https://coderwall.com/p/zpwrsg/add-search-function-to-list-view-in-android
            ShowArticleList(filter);
            this.lastSearchText = filter;
            return true;
        }

        public bool OnQueryTextSubmit(string filter)
        {
            if (this.lastSearchText == filter)
                return true;

            // Filter ggf. mit Adapter, siehe https://coderwall.com/p/zpwrsg/add-search-function-to-list-view-in-android
            ShowArticleList(filter);
            this.lastSearchText = filter;
            return true;
        }

    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/ArticleListActivity.kt
package de.stryi.vorratsuebersicht

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.SearchView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.menu.MenuBuilder
import androidx.appcompat.widget.PopupMenu
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import de.stryi.vorratsuebersicht.ArticleListViewAdapter.ArticleViewHolder
import de.stryi.vorratsuebersicht.database.Database
import de.stryi.vorratsuebersicht.databinding.ArticleListBinding
import de.stryi.vorratsuebersicht.tools.AddToShoppingListDialog
import de.stryi.vorratsuebersicht.tools.CategoryItem
import de.stryi.vorratsuebersicht.tools.Tools
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter


class ArticleListActivity : AppCompatActivity() {

    private var selectArticleOnly = false
    private var category = ""
    private var subCategory = ""
    private var withoutCategory = false
    private var notInStorage = false
    private var notInShoppingList = false
    private val eanCode: String? = null
    private var lastSearchText: String? = ""
    private var specialFilter = 0

    private var listViewState: Parcelable? = null
    private lateinit var binding: ArticleListBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ArticleListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        this.setSupportActionBar(binding.ArticleListAppBar)

        binding.ArticleListAppBar.setNavigationOnClickListener { finish() }
        binding.ArticleListAppBar.overflowIcon?.setTint(getColor(R.color.Application_ActionBar_TextColor))

        this.selectArticleOnly = intent.getBooleanExtra("SelectArticleOnly", false)
        this.notInStorage      = intent.getBooleanExtra("NotInStorage", false)
        this.notInShoppingList = intent.getBooleanExtra("NotInShoppingList", false)

        if (this.selectArticleOnly)
        {
            binding.ArticleListAppBar.title = "Auswahl Artikel"
        }

        // Trennlinie in der Liste
        val divider = DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
        ContextCompat.getDrawable(this, R.drawable.divider)?.let { divider.setDrawable(it) }
        binding.ArticleList.addItemDecoration(divider)
        binding.ArticleList.isClickable = true

        // "Pull to Refresh" Funktionalität
        binding.ArticleListSwipeRefreshLayout.setOnRefreshListener {
            this.showArticleList()
            this.loadCategoryList()
            binding.ArticleListSwipeRefreshLayout.isRefreshing = false
        }

        // Kategorie Auswahl
        this.loadCategoryList()

        binding.ArticleListCategories.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View?, position: Int, id: Long) {
                    val categoryItem = parent.getItemAtPosition(position) as CategoryItem
                    spinnerCategoryItemSelected(position, categoryItem.category, categoryItem.subCategory)
                }

                override fun onNothingSelected(parent: AdapterView<*>) {
                    // Optional: Verhalten, wenn nichts ausgewählt ist
                    spinnerCategoryItemSelected(0, "", "")
                }
            }

        // Spezialfilter entfernen
        binding.ArticleListFilterClear.setOnClickListener {
            this.specialFilter = 0
            this.binding.ArticleListFilter.text = ""
            this.binding.ArticleListFilterBanner.visibility = View.GONE

            this.showArticleList()
            this.loadCategoryList()
        }

        binding.ArticleListFABAdd.setOnClickListener {
            onCreateArticle()
        }

        showArticleList()
    }

    private fun loadCategoryList()
    {
        val categoryList = mutableListOf<CategoryItem?>()
        categoryList.add(CategoryItem(resources.getString(R.string.ArticleList_AllCategories)))
        categoryList.add(CategoryItem(resources.getString(R.string.ArticleList_NoCategories)))
        categoryList.addAll(Database.getCategoryAndSubcategoryNames())

        val dataAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, categoryList)
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

        binding.ArticleListCategories.adapter = dataAdapter
    }

    @SuppressLint("RestrictedApi")
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (menu is MenuBuilder) {
            menu.setOptionalIconsVisible(true)
        }
        menuInflater.inflate(R.menu.article_list_menu, menu)

        val searchItem = menu.findItem(R.id.ArticleList_Menu_Search)
        val searchView = searchItem.actionView as SearchView

        val editText = Tools.findFirstEditText(searchView)

        editText?.setTextColor(getColor(R.color.white))

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                // Optional: Suche abschließen
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                lastSearchText = newText.orEmpty()
                showArticleList()
                return true
            }
        })

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.ArticleList_Menu_Add -> {
                onCreateArticle()
                return true
            }
            R.id.ArticleList_Menu_Filter -> {
                // Spezialfilter auswählen
                this.filterArticleList()
                return true
            }
            R.id.ArticleList_Menu_Share -> {
                shareList()
                return true
            }
            else -> return false
        }
    }

    private fun filterArticleList()
    {
        // Auswahl Spezialfilter
        val actions = resources.getTextArray(R.array.ArticleListeSpecialFilter)

        val adapter = ArrayAdapter(
            this,
            R.layout.dialog_item,
            R.id.DialogItem_Text1,
            actions)

        val builder = AlertDialog.Builder(this, R.style.MyAlertDialogTheme)
        builder.setAdapter(adapter) { _, which ->
            this.specialFilter = which + 1

            binding.ArticleListFilter.text = actions[which]
            binding.ArticleListFilterBanner.visibility = View.VISIBLE

            this.showArticleList()
        }
        builder.show()
    }

    fun showArticleList()
    {
        val articleList = Database.getArticleList(
            this.category,
            this.subCategory,
            this.eanCode,
            this.notInStorage,
            this.notInShoppingList,
            this.withoutCategory,
            this.specialFilter,
            this.lastSearchText)

        val adapter = ArticleListViewAdapter(articleList, this::onOpenArticleDetails)
        adapter.optionSelect =
            { articleId, anchor ->
                this.showOptionPopUp(articleId, anchor)
            }

        binding.ArticleList.layoutManager = LinearLayoutManager(this)
        binding.ArticleList.adapter = adapter

        val status = if (articleList.count() == 1)
            resources.getString(R.string.ArticleListSummary_Position)
        else
            resources.getString(R.string.ArticleListSummary_Positions)

        binding.ArticleListFooter.text = String.format(status, articleList.count())
    }

    fun onCreateArticle()
    {
        listViewState = binding.ArticleList.layoutManager?.onSaveInstanceState()

        val intent = Intent(this, ArticleDetailsActivity::class.java)

        // Zum Voranstellen bei Neuanlage
        intent.putExtra("Category",    this.category)
        intent.putExtra("SubCategory", this.subCategory)
        if (this.selectArticleOnly)
        {
            intent.putExtra("NoStorageQuantity", true)
            intent.putExtra("NoDeleteArticle",   true)
        }

        detailLauncher.launch(intent)
    }

    fun onOpenArticleDetails(articleId: Int)
    {
        if (this.selectArticleOnly)
        {
            val intent = Intent()
            intent.putExtra("ArticleId", articleId)
            setResult(RESULT_OK, intent)
            finish()
            return
        }

        this.showArticleDetails(articleId)
    }

    fun showArticleDetails(articleId: Int) {

        val articleDetails = Intent(this, ArticleDetailsActivity::class.java)
        articleDetails.putExtra("ArticleId", articleId)

        detailLauncher.launch(articleDetails)

        listViewState = binding.ArticleList.layoutManager?.onSaveInstanceState()
    }


    private val detailLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                /*
                var articleId = result.data?.getIntExtra("ArticleId", 0)
                if (articleId == 0)
                    return@registerForActivityResult

                var article = Database.getArticle(articleId!!)

                var adapter = binding.ArticleList.adapter as ArticleListViewAdapter

                adapter.notifyItemChanged(index)
                */

                showArticleList()
                loadCategoryList()

                binding.ArticleList.layoutManager?.onRestoreInstanceState(listViewState)
            }
        }

     fun spinnerCategoryItemSelected(position: Int, categoryText: String, subCategoryText: String)
     {
         val withoutCategoryNew = position == 1

         if ((categoryText != this.category) || (subCategoryText != this.subCategory) || withoutCategoryNew != this.withoutCategory)
         {
             this.category        = categoryText
             this.subCategory     = subCategoryText
             this.withoutCategory = withoutCategoryNew

             this.showArticleList()
         }
     }

    fun shareList()
    {
        if (MainActivity.IsGooglePlayPreLaunchTestMode)
        {
            return
        }

        val list = Database.getArticleList(
            this.category,
            this.subCategory,
            this.eanCode,
            this.notInStorage,
            this.notInShoppingList,
            this.withoutCategory,
            this.specialFilter,
            lastSearchText)

        var text = ""

        for(article in list)
        {
            if (article.heading.isNotEmpty())     text += article.heading + "\n"
            if (article.articleInfo.isNotEmpty()) text += article.articleInfo + "\n"
            if (article.notesText.isNotEmpty())   text += article.notesText + "\n"
            text += "\n"
        }

        text += binding.ArticleListFooter.text

        val now = LocalDateTime.now()

        val subject = String.format("%s - %s",
            resources.getString(R.string.Main_Button_Artikelangaben),
            now.format(DateTimeFormatter.ofPattern("dd.MM.yyyy HH:mm")))

        val intent = Intent()
        intent.action = Intent.ACTION_SEND
        intent.putExtra(Intent.EXTRA_SUBJECT, subject)
        intent.putExtra(Intent.EXTRA_TEXT, text)
        intent.type = "text/plain"

        startActivity(intent)
    }

    @SuppressLint("RestrictedApi")
    fun showOptionPopUp(articleId: Int, anchor: ArticleViewHolder)
    {
        val popupMenu = PopupMenu(this, anchor.option)
        popupMenu.menuInflater.inflate(R.menu.article_list_contextmenu, popupMenu.menu)

        if (popupMenu.menu is MenuBuilder) {
            (popupMenu.menu as MenuBuilder).setOptionalIconsVisible(true)
        }

        popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
            when (menuItem.itemId) {
                R.id.ArticleList_ContextMenu_Lagerbestand -> {
                    val storageDetails = Intent(anchor.itemView.context, StorageItemInventoryActivity::class.java)
                    storageDetails.putExtra("ArticleId", articleId)
                    storageDetailsLauncher.launch(storageDetails)
                    true
                }
                R.id.ArticleList_ContextMenu_AufEinkaufszettel -> {
                    AddToShoppingListDialog.showDialog(
                        anchor.itemView.context as Activity,
                        articleId,
                        anchor.minQuantity,
                        anchor.prefQuantity,
                        { refreshArticleList() }
                    )
                    true
                }
                else -> false
            }
        }
        popupMenu.show()
    }

    private val storageDetailsLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult())
    { result ->
        if (result.resultCode == RESULT_OK)
        {
            this.refreshArticleList()
        }
    }

    private fun refreshArticleList()
    {
        listViewState = binding.ArticleList.layoutManager?.onSaveInstanceState()
        this.showArticleList()
        binding.ArticleList.layoutManager?.onRestoreInstanceState(listViewState)
    }
}